File: pyg4ometry/pygeometry/cpgeant4/solid/Wedge.cpp
#include "Wedge.h"

CSG* CSGMesh::ConstructWedge(double pRMax,double pSPhi,double pDPhi,double halfzlength){
  double d = halfzlength;
  int nslice = 16;
  std::vector<double> x(nslice);
  std::vector<double> y(nslice);
  std::vector<Vertex*> p1(nslice);
  std::vector<Vertex*> p2(nslice);

  for(int i=0;i<nslice;i++){
    double phi = pSPhi + double(i)*(pDPhi-pSPhi)/double(nslice);
    x[i] = pRMax*cos(phi);
    y[i] = pRMax*sin(phi);
    p1[i] = new Vertex(new Vector(x[i],y[i],-d),NULL);
    p2[i] = new Vertex(new Vector(x[i],y[i], d),NULL);
  }

  std::vector<Polygon*> polygons;
  Vertex* vZero1 = new Vertex(new Vector(0.0,0.0,-d),NULL);
  Vertex* vZero2 = new Vertex(new Vector(0.0,0.0, d),NULL);

  for(int i=0;i<nslice-1;i++){
    //top triangle
    std::vector<Vertex*> vtx1(3);
    vtx1[0] = vZero2;vtx1[1] = p2[i];vtx1[2]=p2[i+1];
    polygons.push_back(new Polygon(vtx1,NULL));

    //bottom triangle
    std::vector<Vertex*> vtx2(3);
    vtx2[0] = vZero1;vtx2[1] = p1[i];vtx2[2]=p1[i+1];
    polygons.push_back(new Polygon(vtx2,NULL));

    //end square
    std::vector<Vertex*> vtx3(4);
    vtx3[0] = p1[i];vtx3[1] = p1[i+1];vtx3[2]=p2[i+1];vtx3[3]=p2[i];
    polygons.push_back(new Polygon(vtx3,NULL));
  }
  //first end face
  std::vector<Vertex*> vtxf(4);
  vtxf[0] = vZero1;vtxf[1] = p1[0];vtxf[2] = p2[0];vtxf[3] = vZero2;
  polygons.push_back(new Polygon(vtxf,NULL));

  std::vector<Vertex*> vtxe(4);
  vtxe[0] = vZero1;vtxe[1] = vZero2;vtxe[2] = p2.back();vtxe[3] = p1.back();
  polygons.push_back(new Polygon(vtxe,NULL));

  return CSG::fromPolygons(polygons);
   
}

File: src/pyg4ometry/geant4/solid/Wedge.py
from ... import config as _config

from .SolidBase import SolidBase as _SolidBase
from ...pycsg.core import CSG as _CSG
from ...pycsg.geom import Vertex as _Vertex
from ...pycsg.geom import Vector as _Vector
from ...pycsg.geom import Polygon as _Polygon
import numpy as _np

from copy import deepcopy as _dc

class Wedge(_SolidBase):
    """
    Constructs a *infinite* wedge. Should not be used to construct geant4 geometry.

    :param name: of object in registry
    :type name: str
    :param normal: normal [x,y,z]
    :type normal: tuple
    :param dist: distance from origin to plane
    :type dist: float
    :param zlength: large transverse box size to emulate infinite plane
    :type zlength: float
    """
    def __init__(self, name, pRMax=1000, pSPhi=0, pDPhi=1.5, halfzlength=10000, nslice=None):
        super(Wedge, self).__init__(name, 'InfiniteWedge', None)

        self.pRMax = float(pRMax)
        self.pSPhi = float(pSPhi)
        self.pDPhi = float(pDPhi)
        self.pDz   = float(halfzlength)
        self.nslice = nslice if nslice else _config.SolidDefaults.Wedge.nslice
        self.mesh = None

    def __repr__(self):
        return 'Wedge : '+self.name+' '+str(self.pRMax)+' '+str(self.pSPhi)+' '+str(self.pDPhi)+' '+str(self.pDz)

    def pycsgmesh(self):

        d = self.pDz

        phi = _np.linspace(self.pSPhi, self.pDPhi, self.nslice)
        x  = self.pRMax*_np.cos(phi)
        y  = self.pRMax*_np.sin(phi)

        polygons = []

        vZero1 = _Vertex(_Vector(0,0,-d),None)
        vZero2 = _Vertex(_Vector(0,0, d),None)

        p1     = [_Vertex(_Vector(x[i],y[i],-d),None)  for i in range(0,len(x))]
        p2     = [_Vertex(_Vector(x[i],y[i], d),None)  for i in range(0,len(x))]


        for i in range(0,len(x)):
            if i != len(x)-1:
                # top triangle
                polygons.append(_Polygon([_dc(vZero2),_dc(p2[i]),_dc(p2[i+1])]))
                # bottom triangle
                polygons.append(_Polygon([_dc(vZero1),_dc(p1[i]),_dc(p1[i+1])]))
                # end square
                polygons.append(_Polygon([_dc(p1[i]),_dc(p1[i+1]),_dc(p2[i+1]),_dc(p2[i])]))

        # first end face
        polygons.append(_Polygon([_dc(vZero1),_dc(p1[0]),_dc(p2[0]),_dc(vZero2)]))

        # last end face
        polygons.append(_Polygon([_dc(vZero1),_dc(vZero2),_dc(p2[-1]),_dc(p1[-1])]))

        mesh = _CSG.fromPolygons(polygons)

        return mesh

